### PR TITLE
[JENKINS-30007] let Installable implement NodeSpecific

### PR DESCRIPTION
**Pull request by ndeloof**

### CODE REVIEW

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 1]
```diff
@@ -64,6 +65,9 @@ public FilePath performInstallation(ToolInstallation tool, Node node, TaskListen
             return expected;
         }
 
+        if (inst instanceof NodeSpecific)
+        inst = (Installable) ((NodeSpecific) inst).forNode(node, log);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Please use brackets or tabs at least. The code is very confusing now

**Comment by oleg-nenashev** [comment]
IMO it should also process <code>EnvironmentSpecific</code>, which is being widely used in Tools

**Comment by ndeloof** [comment]
@oleg-nenashev indeed, I can't find a use-case for now, but doesn't cost much to get this in core

**Comment by oleg-nenashev** [comment]
Agreed. :bee: for the current state if you expect the formal review (the reference is missing)

**Comment by ndeloof** [comment]
@reviewbybees

**Comment by jglick** [comment]
I would expect to see the work-in-progress downstream pull request demonstrating use of this change.

**Comment by ndeloof** [comment]
@jglick I'm working on one for nodejs plugin, will publish tomorrow

**Comment by ndeloof** [comment]
see https://github.com/jenkinsci/nodejs-plugin/pull/10 as example use

File: core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java [hunk 2]
```diff
@@ -175,4 +180,16 @@ public String getId() {
          */
         public String url;
     }
+
+    /**
+     * Convenient abstract class to implement a NodeSpecificInstallable based on an existing Installable
+     */
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java]: Needs <code>@since TODO</code>

**Comment by oleg-nenashev** [comment]
I don't see much added value in <code>NodeSpecificInstallable</code>, but I'll be OK with it if it is documented properly

**Comment by ndeloof** [comment]
@oleg-nenashev you can't create an anonymous class to both extend Installable and implement NodeSpecific, so this class make it possible to write https://github.com/ndeloof/nodejs-plugin/commit/14945f01d4b4e4d16f2804f2d305b853f7f6d74e#diff-e73e9ba7758923527f245503d35f133bR88 as an anonymous class. Would ideally be a decorator but Installable is based on public fields.

**Comment by oleg-nenashev** [comment]
I don't like such anonymous classes, but it's just a personal IMHO.

:bee:

**Comment by ndeloof** [comment]
@oleg-nenashev I'd prefer to attach a lambda to an Installable to make it NodeSpecific, but we are not (yet) designing Java 8 APIs :P

**Pull request merged**